File: src/main/kotlin/com/manpower/laboratory/model/entity/Artifact.java
package com.manpower.laboratory.model.entity;


import io.quarkus.hibernate.orm.panache.PanacheEntity;
import io.quarkus.runtime.annotations.RegisterForReflection;

import javax.persistence.CascadeType;
import javax.persistence.Entity;
import javax.persistence.OneToMany;
import java.util.List;

@Entity
@RegisterForReflection
public class Artifact extends CustomPanacheEntity {

    public String artifactId;
    public String groupId;

    @OneToMany(mappedBy = "artifact", cascade = CascadeType.ALL)
    public List<ArtifactVersion> versions;

    public Artifact() {
    }

    public Artifact(String artifactId, String groupId) {
        this.artifactId = artifactId;
        this.groupId = groupId;
    }
}

File: src/main/kotlin/com/manpower/laboratory/model/dto/Artifact.kt
package com.manpower.laboratory.model.dto

import com.manpower.laboratory.model.entity.Artifact
import com.manpower.laboratory.model.entity.ArtifactInstance
import com.manpower.laboratory.model.entity.ArtifactVersion

class ArtifactDto(artifact: Artifact) {
    val artifactId = artifact.artifactId
    val groupId = artifact.groupId
    val versions = artifact.versions.map { ArtifactVersionDto(it) }
}

class ArtifactVersionDto(artifactVersion: ArtifactVersion) {
    val number = artifactVersion.number
    val instances = artifactVersion.artifactInstances.map { ArtifactInstanceDto(it) }
}

class ArtifactInstanceDto(artifactInstance: ArtifactInstance) {
    val environmentType: String = artifactInstance.environmentType.name
}
